### ISSUE TITLE
ext_authz: Consider sending enveloped request to http_service instead of as-is

### ISSUE THREAD

**Issue by acruise** [envoyproxy/envoy#19276]
*Title*: ext_authz: Consider sending enveloped request to http_service instead of as-is

*Description*:
I've been looking into using envoy as a non-transparent, authenticated egress proxy (i.e. using HTTP CONNECT). External Authorization seems like a good way to go, but when I tried to use an `http_service` I saw that the requests were being copied to the ext_authz service as-is, i.e. a CONNECT request from behind the proxy was being sent as a CONNECT request to the ext_authz service.  This seems like a nice symmetry, but in my case, since very few HTTP server frameworks support CONNECT, I couldn't find a good way to parse these requests so they could be authorized.  I wound up switching to gRPC (#19246) but this isn't ideal, because at the moment we have ~0 other use of gRPC in production, so when I hand off the ext_authz service to someone else, they'll have an additional learning curve.

>Describe the desired behavior, what scenario it enables and how it
would be used.

For `CONNECT` (and probably other types of requests), it'd be nice if a future `http_service` external authorization protocol could support wrapping the original request in an envelope (JSON would probably be nice), and send it to the authz service using an identical protocol every time, no matter what the original request looks like. Maybe something like this?:

```
POST /ext-authz/authorize_request HTTP/1.1
Content-Type: application/json
Content-Length: nnnnn

originalRequest: {
  method: POST
  raw_url: https://example.com/foo?bar=baz
  parsed_url: {
    scheme: https
    host: example.com
    path: /foo
    query: {
      bar: baz
    }
  }
  headers: {
    Cookie: "abcd123"
    ...
  }
  body: "<base64 or escaped JSON of original body>"
}


```

**Comment by phlax** [comment]
cc @dio

**Comment by esmet** [comment]
I think you can use something like https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_filters/grpc_http1_reverse_bridge_filter to support a server that still understands protobuf and HTTP 1.1

This isn't the JSON solution you suggested, but it does remove gRPC.

**Comment by acruise** [comment]
> server that still understands protobuf and HTTP 1.1

yeah, that would be better, but still pretty alien. I'm going ahead with gRPC for now so this enhancement isn't a high priority for me anyway. :)
